### PR TITLE
Task-51907 : Add winners list drawer

### PR DESCRIPTION
**Pull request by Rdidmohamedelhadi**

### CODE REVIEW

**Review by GouadriaHanen — APPROVED** [review]
LGTM

File: challenges-api/src/main/java/org/exoplatform/challenges/service/AnnouncementService.java [hunk 1]
```diff
@@ -50,10 +50,19 @@ Announcement createAnnouncement(Announcement announcement, String username) thro
...
                                                                                              ObjectNotFoundException;
 
+  /**
+   * Retrieves number of all Announcements by challengeId.
+   *
+   * @return A {@link Long} number of announcements
+   * @throws ObjectNotFoundException when the challenge identified by its
+   *           technical identifier is not found
+   */
+  Long countAllAnnouncementByChallenge(long challengeId) throws ObjectNotFoundException;
```
Comment (ahamdi) [inline challenges-api/src/main/java/org/exoplatform/challenges/service/AnnouncementService.java]:
```suggestion
  Long countAnnouncementsByChallenge(long challengeId) throws ObjectNotFoundException;
```

File: challenges-api/src/main/java/org/exoplatform/challenges/model/ChallengeRestEntity.java [hunk 2]
```diff
@@ -30,7 +30,18 @@
 
   private List<AnnouncementRestEntity> announcements;
 
+  private Long                         numberAllAnnouncements;
```
Comment (ahamdi) [inline challenges-api/src/main/java/org/exoplatform/challenges/model/ChallengeRestEntity.java]:
```suggestion
  private Long                         announcementsCount;
```

File: challenges-api/src/main/java/org/exoplatform/challenges/model/ChallengeRestEntity.java [hunk 3]
```diff
@@ -30,7 +30,18 @@
...
+    return new ChallengeRestEntity(id,
+                                   title,
+                                   description,
+                                   space,
+                                   startDate,
+                                   endDate,
+                                   userInfo,
+                                   managers,
+                                   announcements,
+                                   numberAllAnnouncements);
```
Comment (ahamdi) [inline challenges-api/src/main/java/org/exoplatform/challenges/model/ChallengeRestEntity.java]:
```suggestion
                                   announcementsCount);
```

File: challenges-services/src/main/java/org/exoplatform/challenges/dao/AnnouncementDAO.java [hunk 4]
```diff
@@ -19,4 +20,15 @@
     return resultList == null ? Collections.emptyList() : resultList;
   }
 
+  public Long countAllAnnouncementByChallenge(Long challengeId) {
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/dao/AnnouncementDAO.java]:
```suggestion
  public Long countAnnouncementsByChallenge(Long challengeId) {
```

File: challenges-services/src/main/java/org/exoplatform/challenges/dao/AnnouncementDAO.java [hunk 5]
```diff
@@ -19,4 +20,15 @@
...
+    TypedQuery<Long> query = getEntityManager().createNamedQuery("Announcement.countAllAnnouncementByChallenge", Long.class);
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/dao/AnnouncementDAO.java]:
```suggestion
    TypedQuery<Long> query = getEntityManager().createNamedQuery("Announcement.countAnnouncementsByChallenge", Long.class);
```

File: challenges-services/src/main/java/org/exoplatform/challenges/entity/AnnouncementEntity.java [hunk 6]
```diff
@@ -11,7 +11,8 @@
 @ExoEntity
 @Table(name = "ANNOUNCEMENT")
 @NamedQueries({
-    @NamedQuery(name = "Announcement.findAllAnnouncementByChallenge", query = "SELECT DISTINCT a FROM Announcement a where a.challenge.id = :challengeId order by a.createdDate desc"), })
+    @NamedQuery(name = "Announcement.findAllAnnouncementByChallenge", query = "SELECT DISTINCT a FROM Announcement a where a.challenge.id = :challengeId order by a.createdDate desc"),
+    @NamedQuery(name = "Announcement.countAllAnnouncementByChallenge", query = "SELECT COUNT(us) FROM Announcement a LEFT JOIN a.assignee us where a.challenge.id = :challengeId"), })
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/entity/AnnouncementEntity.java]:
```suggestion
    @NamedQuery(name = "Announcement.countAnnouncementsByChallenge", query = "SELECT COUNT(us) FROM Announcement a LEFT JOIN a.assignee us where a.challenge.id = :challengeId"), })
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 7]
```diff
@@ -87,6 +85,18 @@ public Announcement createAnnouncement(Announcement announcement, String current
     return announcementStorage.findAllAnnouncementByChallenge(challengeId, offset, limit);
   }
 
+  @Override
+  public Long countAllAnnouncementByChallenge(long challengeId) throws ObjectNotFoundException {
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]:
```suggestion
  public Long countAnnouncementsByChallenge(long challengeId) throws ObjectNotFoundException {
```

File: challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java [hunk 8]
```diff
@@ -87,6 +85,18 @@ public Announcement createAnnouncement(Announcement announcement, String current
...
+    if (challengeId <= 0) {
+      throw new IllegalArgumentException("Challenge id has to be positive integer");
+    }
+    Challenge challenge = challengeStorage.getChallengeById(challengeId);
+    if (challenge == null) {
+      throw new ObjectNotFoundException("challenge does not exist");
+    }
+    return announcementStorage.countAllAnnouncementByChallenge(challengeId);
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/service/AnnouncementServiceImpl.java]:
```suggestion
    return announcementStorage.countAnnouncementsByChallenge(challengeId);
```

File: challenges-services/src/main/java/org/exoplatform/challenges/storage/AnnouncementStorage.java [hunk 9]
```diff
@@ -46,4 +46,9 @@ public Announcement getAnnouncementById(long announcementId) {
     return EntityMapper.fromAnnouncementEntities(announcementEntities);
   }
 
+  public Long countAllAnnouncementByChallenge(Long challengeId) {
```
Comment (ahamdi) [inline challenges-services/src/main/java/org/exoplatform/challenges/storage/AnnouncementStorage.java]:
```suggestion
  public Long countAnnouncementsByChallenge(Long challengeId) {
```

**Review by ahamdi — COMMENTED** [review]
I added some comments and proposed changes

**Comment by Rdidmohamedelhadi** [comment]
> I added some comments and proposed changes

yes it's a good proposal ! thanks @ahamdi

**Review by ahamdi — APPROVED** [review]

**Pull request merged**